From b0514b09d7a1f5f3086fdfe569218d134c6dd185 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Tue, 3 Sep 2024 15:20:29 +0800
Subject: [PATCH 130/149] mtk: hostapd: do reconnect if receiving CSA during
 association

When receiving CSA event, wpa_s notifies hostapd so that the whole
device can switch to new channel together. However, if the CSA is
received before the association is finished, wpa_s fails to notifies
hostapd since the AP interfaces are disabled. Later the AP interface
setup may fail because wpa_s start up AP interface by old channel
information, which fails kernel's iface_combination check.

This commit triggers a reconnection if CSA is received during
association.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 wpa_supplicant/ucode.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/wpa_supplicant/ucode.c b/wpa_supplicant/ucode.c
index c632349e1..813514ea6 100644
--- a/wpa_supplicant/ucode.c
+++ b/wpa_supplicant/ucode.c
@@ -124,6 +124,14 @@ void wpas_ucode_event(struct wpa_supplicant *wpa_s, int event, union wpa_event_d
 		int is_24ghz;
 		u8 channel;
 
+		if (wpa_s->wpa_state < WPA_COMPLETED) {
+			wpa_printf(MSG_INFO,
+				   "ucode: mtk: receive channel switch event"
+				   "before association finished, reconnect\n");
+			wpa_supplicant_reconnect(wpa_s);
+			return;
+		}
+
 		control_freq = data->ch_switch.freq;
 		center_freq1 = data->ch_switch.cf1;
 
-- 
2.45.2

